crates/pairwise_nostd_internal/src/parallel.rs

```diff
@@ -0,0 +1,724 @@
+///   2. Then, these separate `accum_state` should be combined (in a "nested
+///      reduction") so that a single member knows the total contribution
+///      from the current unit of work.
+///   3. Finally, that single member will use the total contribution to update
+///      the accum_state from the appropriate statepack bin.
+///
+/// when [`reductionspec::nested_reduce`] is `false`, then:
+/// - we don't know of any relationship between the data elements and the
+///   associated bin indices in a "unit-of-work." This really restricts the
+///   level of parallelism, we can achieve.
+/// - The best we can do is adopt a "batching strategy," that has the
+///   has the team members use [`get_datum_index_pair`] to pre-generate a
+///   batch of (datum, bin-index) pairs from the data source. (the batch of
+///   entries is to be sequentially processed)
+/// - the intention is to:
+///    1. have the team members collectively call [`get_datum_index_pair`].
+///       Each member pre-generates a (datum, bin-index) pair from the data
+///       source, represented as a [`BinnedDatum`] instance
+///    2. Then the [`BinnedDatum`] instances should be gathered into a
+///       collection-pad accessible in the memory of a single team member
+///    3. have that team member process the batch of values in
+///       [`BinnedDatum`] values to sequentially update the binned
+///       statepack
+///
+/// # Development Note
+/// An important premise is that you do error-checking while constructing
+/// structs, and you design the logic such that you don't have to do **any**
+/// error-handling in this trait's methods.
+///
+/// # Ideas for Improvement:
+/// The use of `ReductionSpec::NESTED_REDUCE` to determine which methods should
+/// be used is suboptimal. While I'm not in a huge rush to do this, this is
+/// sign that we should break up this trait.
+///
+/// If we have a module `reduction`, I could imagine that we might define
+/// traits with names like
+/// - `reduction::Nested`
+/// - `reduction::BinnedDataBatching`
+///
+/// where each trait is a superset of a third trait (maybe `ReductionInfo`?)
+/// that declares common methods. (If we do this, then we would need to split
+/// apart [`fill_single_team_statepack`] into separate functions)
+pub trait ReductionSpec {
+    // this trait's associated items are split into 2 parts.
+    //   1. the associated supported by all trait implementers
+    //   2. associated items conditionally supported by trait implementers
+    // TODO: Figure out how to better express the conditionally implemented
+    //       associated items (maybe use multiple traits?)
+
+    // Associated items that are always used
+    // -------------------------------------
+    type AccumulatorType: Accumulator;
+
+    /// return a reference to the accumulator
+    fn get_accum(&self) -> &Self::AccumulatorType;
+
+    /// The number of bins in this reduction.
+    fn n_bins(&self) -> usize;
```

Comment:
```suggestion
    /// The number of distance bins in this reduction.
    fn n_bins(&self) -> usize;
```

I can't remember, did we decide on a different terminology?

Reply:
Previously, we were calling it distance bins and separation bins pretty interchangeably. I think I like the idea of referring to these generically as bins because it makes the reduction machinery easier to reason about in isolation from structure functions and correlation functions.

- I think the original concern was that it might be confusing with the `Histogram` accumulator. But, I think we can work around that by saying the Histogram Accumulator has "buckets" or "counters."
- If you have strong thoughts on this, I could be swayed